corehq/apps/smsbillables/models.py

```diff
@@ -358,62 +338,84 @@ def create(cls, message_log, multipart_count=1):
 
     @classmethod
-    def _get_multipart_count(cls, backend_api_id, backend_instance, backend_message_id, multipart_count):
-        if backend_api_id == SQLTwilioBackend.get_api_id():
-            twilio_message = get_twilio_message(backend_instance, backend_message_id)
-            if twilio_message.num_segments is not None:
-                return int(twilio_message.num_segments)
-            else:
-                raise RetryBillableTaskException("backend_message_id=%s" % backend_message_id)
-        elif backend_api_id == InfobipBackend.get_api_id():
-            infobip_message = get_infobip_message(backend_instance, backend_message_id)
-            segments = infobip_message['messageCount'] \
-                if 'messageCount' in infobip_message else infobip_message['smsCount']
-            if segments is not None:
-                return int(segments)
-            else:
-                raise RetryBillableTaskException("backend_message_id=%s" % backend_message_id)
-        else:
-            return multipart_count
-
-    @classmethod
-    def _get_gateway_fee(cls, backend_api_id, backend_instance,
+    def _get_gateway_fee(cls, backend_api_id, backend_id,
                          phone_number, direction, couch_id, backend_message_id, domain):
         country_code, national_number = get_country_code_and_national_number(phone_number)
-        is_gateway_billable = backend_instance is None or _sms_backend_is_global(
-            backend_instance) or toggles.ENABLE_INCLUDE_SMS_GATEWAY_CHARGING.enabled(domain)
+
+        backend_instance = None
+        if backend_id is not None:
+            backend_instance = SQLMobileBackend.load(
+                backend_id,
+                api_id=backend_api_id,
+                is_couch_id=True,
+                include_deleted=True,
+            )
+
+        is_gateway_billable = (
+            backend_id is None
+            or backend_instance.is_global
+            or toggles.ENABLE_INCLUDE_SMS_GATEWAY_CHARGING.enabled(domain)
+        )
+
+        direct_gateway_fee = gateway_fee = multipart_count = conversion_rate = None
 
         if is_gateway_billable:
-            if backend_api_id in [SQLTwilioBackend.get_api_id(), InfobipBackend.get_api_id()]:
-                provider_charges = cls._get_provider_charges(
-                    backend_message_id, backend_instance, direction, couch_id, backend_api_id
-                )
-                gateway_fee = provider_charges.gateway_fee
-                direct_gateway_fee = provider_charges.provider_gateway_fee
+            if backend_instance and backend_instance.using_api_to_get_fees:
+                if backend_message_id:
+                    direct_gateway_fee, multipart_count = \
+                        cls.get_charge_details_through_api(backend_instance, backend_message_id)
+
+                    gateway_fee = SmsGatewayFee.get_by_criteria(
+                        backend_api_id,
+                        direction,
+                    )
```

Comment:
@nemanja1040 is there a reason the `gateway_fee` needs to be created here with limited criteria input? Or could we potentially remove this block and always have the below code that gets an `SmsGatewayFee` with all criteria specified execute? Let me know if I need to clarify my question. Thanks.

Reply:
Hey @gherceg, I believe you are right. Here is my suggestion:

    @classmethod
    def _get_gateway_fee(cls, backend_api_id, backend_id,
                         phone_number, direction, couch_id, backend_message_id, domain):
        country_code, national_number = get_country_code_and_national_number(phone_number)
        backend_instance = SQLMobileBackend.load(
            backend_id,
            api_id=backend_api_id,
            is_couch_id=True,
            include_deleted=True,
        )

        is_gateway_billable = backend_id is None\
                              or backend_instance.is_global\
                              or toggles.ENABLE_INCLUDE_SMS_GATEWAY_CHARGING.enabled(domain)

        direct_gateway_fee = gateway_fee = multipart_count = conversion_rate = None

        if is_gateway_billable:
            if backend_message_id:
                if backend_instance.using_api_to_get_fees:
                    direct_gateway_fee, multipart_count = \
                        cls.get_charge_details_through_api(backend_instance, backend_message_id)

                gateway_fee = SmsGatewayFee.get_by_criteria(
                    backend_api_id,
                    direction,
                    backend_instance=backend_id,
                    country_code=country_code,
                    national_number=national_number,
                )

            else:
                log_smsbillables_error(
                    "Could not create gateway fee for message %s: no backend_message_id" % couch_id
                )

            if gateway_fee:
                conversion_rate = cls.get_conversion_rate(gateway_fee)
            else:
                log_smsbillables_error(
                    "No matching gateway fee criteria for SMS %s" % couch_id
                )
        return _ProviderChargeInfo(
            direct_gateway_fee,
            gateway_fee,
            multipart_count,
            conversion_rate
        )